lib/TPP/RewriteToBatchReduceGemm.cpp

```diff
@@ -180,8 +180,8 @@ getSlicedOperands(OpBuilder &builder, Location loc, ValueRange localIvs,
   SmallVector<Value> slicedOperands;
   for (OpOperand *operand : linalgOp.getDpsInputOperands()) {
     FailureOr<Value> slicedOperand;
-    AffineMap map = linalgOp.getMatchingIndexingMap(operand);
-    if (isVNNILoop && map.getNumResults() == 5) {
+    // In VNNI layout the second operand has size '4'.
+    if (isVNNILoop && operand->getOperandNumber() == 1) {
```

Comment:
Getting the order of the operands is even flakier than the number of results... There must be a better way of doing this?

Reply:
Why? Looking at the op documentation the second operand must be 4d.

Reply:
What matters isn't the documentation, but the match that leads here.

If I'm reading correctly, `isLinalgOpVNNI` can match both ranks 4 and 5, which seems to be the only check before calling this function.

In what cases it is 4 or 5 and can we have tests for both?

Reply:
Kavitha, will check tomorrow. My understanding was only 4.

Reply:
https://github.com/plaidml/tpp-mlir/blob/main/lib/TPP/RewriteToBatchReduceGemm.cpp#L213

Reply:
The whole looks very C&P to me, why do we have 2 routines which are close to being identical for mapping to BRGEMM and VNNI BRGEMM. 
The line 213 @rengolin  is referring to should be removed, this is what I recently discussed with @KavithaTipturMadhu . The check for the rank of the inner dimension of the second operand being equal to the VNNI blocking factor should go into "CheckVNNIStruture".

I also had a look at this: https://github.com/plaidml/tpp-mlir/blob/9311ad0c727c83134c0b894bc608d9d7b6f1ba45/lib/TPP/Dialect/Tpp/TppUtils.cpp#L92 I think we should not name the MatMulBody but rather FMA or Linear.

Reply:
I have renamed hasMatmulBody to hasFMABody in PR https://github.com/plaidml/tpp-mlir/pull/387